gqcp/src/ONVBasis/SpinUnresolvedOperatorString.cpp

```diff
@@ -0,0 +1,130 @@
+    // Since the operator string will not be zero in this case, we must return false.
+    return (iterator != index_vector.end());
+}
+
+
+/*
+*  MARK: Public methods
+*/
+
+/**
+ *  Retrieve the phase factor after sorting the `SpinUnresolvedOperatorString`.
+ * 
+ *  Note: Please refer to method `sort()` to perform the actual sorting.
+ * 
+ *  @return The phase factor after sorting the operator string.
+ */
+int SpinUnresolvedOperatorString::phaseFactorAfterSorting() {
+
+    auto index_vector = this->operatorIndices();
+    const auto number_of_occupied_spinors = index_vector.size();
+    int phase_factor = 1;
+
+    for (int i = 0; i < number_of_occupied_spinors; ++i) {
+
+        for (int j = i; j >= 0; --j) {
+            if (index_vector[j] > index_vector[i]) {
+                phase_factor *= -1;
+            }
+        }
+    }
+    return phase_factor;
+}
+
+
+/**
+ *  Sort the operator string (in-place) in ascending order and adjust its phase factor. If two indices are equal, the phase factor may not be correct but it does not really matter since its effect on a wavefunction is zero.
+ */
+void SpinUnresolvedOperatorString::sort() {
+
+    const auto phase_factor = this->phaseFactorAfterSorting();
+
+    std::sort(this->indices.begin(), this->indices.end());
+    this->p *= phase_factor;
+}
+
+
+/**
+ *  Decomposition of the `SpinUnresolvedOperatorString` into two new operator strings: a system and an environment.
+ * 
+ *  @param partition    The decomposition of the operator string into a system (denoted by 'I') and an environment (denoted by 'J').
+ *  
+ *  For example: Operator string "a1a2a4a0a3" is decomposed into system "a1a4a0" and environment "a2a3" by the partition {'I', 'J', 'I', 'I', 'J'}.
+ */
+std::vector<SpinUnresolvedOperatorString> SpinUnresolvedOperatorString::schmidtDecomposition(const std::vector<char>& partition) {
+
+    if (partition.size() != this->operatorIndices().size()) {
+        throw std::invalid_argument("SpinUnresolvedOperatorString::schmidtDecomposition(const std::vector<char>& partition): The size of the partition does not match the number of operators in the operator string.");
+    }
+
+    const auto& index_vector = this->operatorIndices();
+
+    std::vector<size_t> index_vector_I;
+    std::vector<size_t> index_vector_J;
+    int phase_factor = this->phaseFactor();
+
+    for (int i = 0; i < index_vector.size(); ++i) {
+
+        if (partition[i] == 'I') {
+            for (int j = i; j >= 0; --j) {
+                if (partition[j] == 'J') {
+                    phase_factor *= -1;
+                }
+            }
+            index_vector_I.push_back(index_vector[i]);
+        } else {
+            index_vector_J.push_back(index_vector[i]);
+        }
+    }
+
+    const SpinUnresolvedOperatorString operator_string_I {index_vector_I, phase_factor};
+    const SpinUnresolvedOperatorString operator_string_J {index_vector_J, 1};
+    return {operator_string_I, operator_string_J};
+}
```

Comment:
I would in general add some more documentation to this function. The loops and if statements can be very tricky to understand for new developers when there is no explanation provided as to what is being done.